plugins/design.cpp

```diff
@@ -0,0 +1,271 @@
+  bool is_extra = type == "extra_point";
+  bool is_in_shape = has_point(x, y);
+  auto mouse_pos = Gui::getMousePos();
+  bool mouse_over = mouse_pos.x == x && mouse_pos.y == y;
+
+  uint32_t pen_key =
+      gen_pen_key(n, s, e, w, is_drag_point, mouse_over, is_in_shape, is_extra);
+
+  if (CURSORS_MAP.count({n, w, e, s}) > 0 && has_point(x,y)) {
+    arr[x][y].cursor_coords = CURSORS_MAP.at({n, w, e, s});
+  }
+
+  if (PENS.find(pen_key) == PENS.end()) {
+    std::pair<int, int> cursor{-1, -1};
+
+    if (type != "") {
+      return make_pen(CURSORS_MAP.at({n, w, e, s}), is_drag_point, mouse_over,
+                      is_in_shape, is_extra);
+    }
+
+    if (CURSORS_MAP.count({n, w, e, s}) > 0) {
+      PENS.emplace(pen_key,
+                   make_pen(CURSORS_MAP.at({n, w, e, s}), is_drag_point,
+                            mouse_over, is_in_shape, is_extra));
+      if (type == "" && has_point(x,y)) {
+        arr[x][y].penKey = pen_key;
+      }
+    }
+  }
+
+  // DEBUG(status).print("not cached lmao\n");
+  return PENS.at(pen_key);
+}
+
+static int design_load_shape(lua_State *L) {
+  if (lua_istable(L, -1)) {
+    lua_pushnil(L);
+    while (lua_next(L, -2) != 0) {
+      int x = lua_tointeger(L, -2);
+
+      if (lua_istable(L, -1)) {
+        lua_pushnil(L);
+        while (lua_next(L, -2) != 0) {
+          int y = lua_tointeger(L, -2);
+          bool value = lua_toboolean(L, -1);
+
+          if (value) {
+            arr[x][y] = DrawingPoint();
+          }
+          lua_pop(L, 1);
+        }
+      }
+      lua_pop(L, 1);
+    }
+  }
+
+  return 0;
+}
+
+static int design_clear_shape(lua_State *L) {
+  arr.clear();
+
+  return 0;
+}
+
+static int design_draw_shape(lua_State *L) {
+  if (arr.size() == 0) {
+    design_load_shape(L);
+  }
+
+  for (auto x : arr) {
+    for (auto y : x.second) {
+      Screen::Pen cur_tile = Screen::readTile(x.first, y.first, true);
+      Screen::Pen pen = get_pen(x.first, y.first, arr);
+      cur_tile.tile = pen.tile;
+      Screen::paintTile(cur_tile, x.first - *window_x, y.first - *window_y,
+                        true);
+    }
+  }
+
+  return 0;
+}
+
+static int design_draw_points(lua_State *L) {
+  if (lua_istable(L, -1)) {
+    const char *str;
+    lua_rawgeti(L, -1, 2);
+    str = lua_tostring(L, -1);
+    lua_pop(L, 1);
+
+    lua_rawgeti(L, -1, 1);
+    int n = luaL_len(L, -1);
+    for (int i = 1; i <= n; i++) {
+      lua_rawgeti(L, -1, i);
+      int x, y;
+      lua_getfield(L, -1, "y");
+      y = lua_tointeger(L, -1);
+      lua_getfield(L, -2, "x");
+      x = lua_tointeger(L, -1);
+      lua_pop(L, 3);
+
+      Screen::Pen cur_tile = Screen::readTile(x, y, true);
+      Screen::Pen pen = get_pen(x, y, arr, str);
+      cur_tile.tile = pen.tile;
+      Screen::paintTile(cur_tile, x - *window_x, y - *window_y, true);
+    }
+    lua_pop(L, 1);
+  }
+
+  return 0;
+}
+
+DFHACK_PLUGIN_LUA_COMMANDS{DFHACK_LUA_COMMAND(design_draw_shape),
```

Comment:
Not for this PR, but future, what would be the best way to have gui/blueprint and `gui/quickfort` use this API? Would it be a good idea to move the rendering logic from the `gui/dwarfmode.lua` library to this plugin?

Reply:
Yeah I've kept that in mind, I wanted to get a smaller scope solution working first, then maybe it can be factored into a more general feature.